torch/csrc/distributed/rpc/request_callback.cpp

```diff
@@ -28,16 +28,24 @@ struct ClearAutogradContextGuard {
 
 } // anonymous namespace
 
-Message RequestCallback::operator()(Message& request) const {
+std::shared_ptr<FutureMessage> RequestCallback::operator()(
+    Message& request) const {
   // For a recv thread, current context id should be invalid outside
   // processMessage().
   ClearAutogradContextGuard guard;
   try {
-    return processMessage(request);
+    // do we need to run clear() in the future invocation?
+    auto ret = processMessage(request);
+    if (!ret->completed()) {
```

Comment:
If I understand this PR correctly, `ret->completed()` should always be true, right? Since when we wrap a message into a Future, we set the completed flag in the constructor. Is there an example of how this change can be used to accommodate a request that should be completed asynchronously?

Reply:
You're right, so I went ahead and added support for BACKWARD_AUTOGRAD_REQ in this change.

It seemed to be mostly a matter of making DistEngine::executeSendFunction() and the underlying clearAndWaitForOutstandingRpcs() return a future.